File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/ContextExternalSources.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Pure-JDK normalization for context sources discovered outside the IDE SDK.
 * MCP resources deliberately remain catalog metadata: callers must use the
 * separately governed read-resource tool before any resource body is loaded.
 */
public final class ContextExternalSources {
    public static final int MAX_MCP_RESOURCES = 20;

    private ContextExternalSources() {}

    @SuppressWarnings("unchecked")
    public static List<Map<String, Object>> parseMcpResources(
            String json, String capturedAt) {
        Object value;
        try {
            value = MiniJson.parse(json == null ? "" : json);
        } catch (RuntimeException error) {
            return List.of();
        }
        if (!(value instanceof List)) return List.of();
        List<?> resources = (List<?>) value;
        List<Map<String, Object>> candidates = new ArrayList<Map<String, Object>>();
        int limit = Math.min(resources.size(), MAX_MCP_RESOURCES);
        for (int index = 0; index < limit; index++) {
            Object item = resources.get(index);
            if (!(item instanceof Map)) continue;
            Map<String, Object> resource = (Map<String, Object>) item;
            String uri = bounded(resource.get("uri"), 512);
            String server = bounded(resource.get("server"), 128);
            if (uri.isEmpty() || server.isEmpty()) continue;
            String nameValue = bounded(resource.get("name"), 160);
            String descriptionValue = bounded(resource.get("description"), 512);
            String mimeTypeValue = bounded(resource.get("mimeType"), 128);
            String name = nameValue.isEmpty() ? null : nameValue;
            String description = descriptionValue.isEmpty()
                    ? null : descriptionValue;
            String mimeType = mimeTypeValue.isEmpty() ? null : mimeTypeValue;

            Map<String, Object> metadata = new LinkedHashMap<String, Object>();
            metadata.put("server", server);
            metadata.put("uri", uri);
            metadata.put("name", name);
            metadata.put("description", description);
            metadata.put("mimeType", mimeType);

            Map<String, Object> freshness = new LinkedHashMap<String, Object>();
            freshness.put("state", "connected-catalog");
            freshness.put("capturedAt",
                    capturedAt == null ? null : bounded(capturedAt, 64));

            Map<String, Object> candidate = new LinkedHashMap<String, Object>();
            candidate.put("kind", "mcp-resource");
            candidate.put("label", "MCP resource: " + (name == null ? uri : name));
            candidate.put("source", "mcp:" + server);
            candidate.put("identity", uri);
            candidate.put("content", MiniJson.stringify(metadata));
            candidate.put("range", null);
            candidate.put("freshness", freshness);
            candidate.put("autoReason",
                    "resource advertised by a connected MCP server");
            candidate.put("refreshable", Boolean.TRUE);
            candidates.add(candidate);
        }
        candidates.sort(Comparator
                .comparing((Map<String, Object> candidate) ->
                        String.valueOf(candidate.get("source")))
                .thenComparing(candidate ->
                        String.valueOf(candidate.get("identity"))));
        return candidates;
    }

    private static String bounded(Object value, int limit) {
        if (value == null) return "";
        String text = String.valueOf(value).trim();
        return text.length() <= limit ? text : text.substring(0, limit);
    }
}

File: packages/vscode-extension/src/context-external-sources.js
"use strict";

const MAX_MCP_RESOURCES = 20;
const DEFAULT_CACHE_MS = 30_000;

function bounded(value, limit) {
  return String(value ?? "")
    .trim()
    .slice(0, limit);
}

function compareText(left, right) {
  if (left === right) return 0;
  return left < right ? -1 : 1;
}

function parseMcpResourceCandidates(value, capturedAt = null) {
  let resources = value;
  if (typeof value === "string") {
    try {
      resources = JSON.parse(value);
    } catch {
      return [];
    }
  }
  if (!Array.isArray(resources)) return [];
  return resources
    .slice(0, MAX_MCP_RESOURCES)
    .map((resource) => {
      if (!resource || typeof resource !== "object") return null;
      const uri = bounded(resource.uri, 512);
      const server = bounded(resource.server, 128);
      if (!uri || !server) return null;
      const name = bounded(resource.name, 160) || null;
      const description = bounded(resource.description, 512) || null;
      const mimeType = bounded(resource.mimeType, 128) || null;
      return {
        kind: "mcp-resource",
        label: `MCP resource: ${name || uri}`,
        source: `mcp:${server}`,
        identity: uri,
        content: JSON.stringify({
          server,
          uri,
          name,
          description,
          mimeType,
        }),
        range: null,
        freshness: {
          state: "connected-catalog",
          capturedAt:
            typeof capturedAt === "string" ? capturedAt.slice(0, 64) : null,
        },
        autoReason: "resource advertised by a connected MCP server",
        refreshable: true,
      };
    })
    .filter(Boolean)
    .sort(
      (left, right) =>
        compareText(left.source, right.source) ||
        compareText(left.identity, right.identity),
    );
}

function createMcpResourceCandidateProvider({
  runCliText,
  getCommand,
  getCwd,
  now = Date.now,
  cacheMs = DEFAULT_CACHE_MS,
} = {}) {
  let cached = [];
  let expiresAt = 0;
  let inFlight = null;
  return async function getMcpResourceCandidates() {
    const current = Number(now()) || 0;
    if (current < expiresAt) return cached;
    if (inFlight) return inFlight;
    inFlight = Promise.resolve()
      .then(() =>
        runCliText({
          command: getCommand(),
          args: ["mcp", "resources", "--json"],
          cwd: getCwd(),
          timeoutMs: 5_000,
        }),
      )
      .then((text) =>
        parseMcpResourceCandidates(
          text,
          new Date(Number(now()) || 0).toISOString(),
        ),
      )
      .catch(() => [])
      .then((next) => {
        cached = next;
        expiresAt =
          (Number(now()) || 0) + Math.max(1_000, Number(cacheMs) || 0);
        return cached;
      })
      .finally(() => {
        inFlight = null;
      });
    return inFlight;
  };
}

module.exports = {
  DEFAULT_CACHE_MS,
  createMcpResourceCandidateProvider,
  parseMcpResourceCandidates,
};
